src/Agent.Worker/Container/ContainerInfo.cs

```diff
@@ -29,32 +33,85 @@ public ContainerInfo(IHostContext hostContext, Pipelines.ContainerResource conta
+            this.ContainerCommand = container.Properties.Get<string>("command", defaultValue: "");
+            this.IsJobContainer = isJobContainer;
 
-#if OS_WINDOWS            
+#if OS_WINDOWS
             _pathMappings[hostContext.GetDirectory(WellKnownDirectory.Tools)] = "C:\\__t"; // Tool cache folder may come from ENV, so we need a unique folder to avoid collision
             _pathMappings[hostContext.GetDirectory(WellKnownDirectory.Work)] = "C:\\__w";
             _pathMappings[hostContext.GetDirectory(WellKnownDirectory.Root)] = "C:\\__a";
+            // add -v '\\.\pipe\docker_engine:\\.\pipe\docker_engine' when they are available (17.09)
 #else
             _pathMappings[hostContext.GetDirectory(WellKnownDirectory.Tools)] = "/__t"; // Tool cache folder may come from ENV, so we need a unique folder to avoid collision
             _pathMappings[hostContext.GetDirectory(WellKnownDirectory.Work)] = "/__w";
             _pathMappings[hostContext.GetDirectory(WellKnownDirectory.Root)] = "/__a";
-#endif            
+            if (this.IsJobContainer)
+            {
+                this.MountVolumes.Add(new MountVolume("/var/run/docker.sock", "/var/run/docker.sock"));
+            }
+#endif
+            if (container.Ports?.Count > 0)
+            {
+                foreach (var port in container.Ports)
+                {
+                    UserPortMappings.Add(port, port);
+                }
+            }
+            if (container.Volumes?.Count > 0)
+            {
+                foreach (var volume in container.Volumes)
+                {
+                    UserMountVolumes.Add(volume, volume);
+                }
+            }
         }
 
         public string ContainerId { get; set; }
         public string ContainerDisplayName { get; private set; }
         public string ContainerNetwork { get; set; }
+        public string ContainerNetworkAlias { get; set; }
         public string ContainerImage { get; set; }
         public string ContainerName { get; set; }
+        public string ContainerCommand { get; set; }
         public string ContainerBringNodePath { get; set; }
         public Guid ContainerRegistryEndpoint { get; private set; }
         public string ContainerCreateOptions { get; private set; }
         public bool SkipContainerImagePull { get; private set; }
-        public IDictionary<string, string> ContainerEnvironmentVariables { get; private set; }
 #if !OS_WINDOWS
         public string CurrentUserName { get; set; }
         public string CurrentUserId { get; set; }
 #endif
+        public bool IsJobContainer { get; set; }
+
+        public IDictionary<string, string> ContainerEnvironmentVariables
+        {
+            get
+            {
+                if (_environmentVariables == null)
+                {
+                    _environmentVariables = new Dictionary<string, string>();
+                }
+
+                return _environmentVariables;
+            }
+            private set
```

Comment:
if you made the constructor change, you can get rid of the `private set`

Reply:
Done